router/src/main.rs

```diff
@@ -343,42 +344,78 @@ async fn main() -> Result<(), RouterError> {
+        .map(|(key, val)| (key.parse::<u32>().unwrap(), val[1..].parse::<u32>().unwrap())).collect();
+        let (mtt, mbs) = tokens_2_bs.iter().next_back().unwrap();
+        if *mtt != (max_total_tokens as u32) {
+            return Err(RouterError::ArgumentValidation(format!("`max_total_tokens` and `max_batch_size` must the same as last item in the 'MAX_TOTAL_TOKENS_2_MAX_BATCH_SIZE_LIST'. Given: {max_total_tokens}, ... and {max_total_tokens_2_bs_pair_str}")));
         }
-        // Flash attention models return their max supported total tokens
-        Some(max_supported_batch_total_tokens) => {
-            // Warn if user added his own max-batch-total-tokens as we will ignore it
-            if max_batch_total_tokens.is_some() {
-                tracing::warn!(
-                    "`--max-batch-total-tokens` is deprecated for Flash \
-                        Attention models."
-                );
-                tracing::warn!(
-                    "Inferred max batch total tokens: {max_supported_batch_total_tokens}"
-                );
-            }
-            if max_total_tokens as u32 > max_supported_batch_total_tokens {
-                return Err(RouterError::ArgumentValidation(format!("`max_total_tokens` must be <= `max_batch_total_tokens`. Given: {max_total_tokens} and {max_supported_batch_total_tokens}")));
+
+        if let Some(max_batch_size) = max_batch_size {
+            if *mbs != (max_batch_size as u32) {
+                return Err(RouterError::ArgumentValidation(format!("`max_total_tokens` and `max_batch_size` must the same as last item in the 'MAX_TOTAL_TOKENS_2_MAX_BATCH_SIZE_LIST'. Given: {max_total_tokens}, {max_batch_size} and {max_total_tokens_2_bs_pair_str}")));
             }
+        }
 
-            max_supported_batch_total_tokens
+    } else {
+        // The 32 will be replaced by max_batch_size in the later
+        tokens_2_bs.insert(max_total_tokens as u32, 32);
+    }
+
+    let mut max_total_tokens_vec: Vec<u32> = Vec::new();
+    for (k, v) in tokens_2_bs{
+        let mtt = k;
+        let mut mit = mtt - 1;
+        let mut mbs: Option<usize> = Some(v as usize);
+        if max_total_tokens_2_bs_pair_str.trim().is_empty() {
+            mit = max_input_tokens as u32;
+            mbs = max_batch_size;
         }
-    };
+
+        let max_supported_batch_total_tokens = match sharded_client
+            .warmup(
+                mit,
+                max_batch_prefill_tokens,
+                mtt as u32,
+                mbs,
+            )
+            .await
+            .map_err(RouterError::Warmup)?
+        {
+            // Older models do not support automatic max-batch-total-tokens
+            None => {
+                let max_batch_total_tokens = max_batch_total_tokens
+                    .unwrap_or(16000.max((max_total_tokens as u32).max(max_batch_prefill_tokens)));
+                tracing::warn!("Model does not support automatic max batch total tokens");
```

Comment:
@donzh I tried `meta-llama/Llama-2-7b-chat-hf` model with adaptive batching and I see this warning in logs, 
`2024-08-06T01:18:30.037184Z  WARN text_generation_router: router/src/main.rs:393: Model does not support automatic max batch total tokens`

any specific reason this model doesn't support adaptive batching ?, which models are considered new models ?
do we have list of models that support adaptive batching ?

Reply:
@donzh , please provide command for TGI benchmark run

Reply:
> @donzh I tried `meta-llama/Llama-2-7b-chat-hf` model with adaptive batching and I see this warning in logs, `2024-08-06T01:18:30.037184Z WARN text_generation_router: router/src/main.rs:393: Model does not support automatic max batch total tokens`
> 
> any specific reason this model doesn't support adaptive batching ?, which models are considered new models ? do we have list of models that support adaptive batching ?

This warning should not prevent you from starting the TGI server. If you don't use adaptive batching the warning exists.
We test the Llama-3-8B, Mixtral and many other models. I think adaptive batching is not related with model. If a model is supported by TGI it should be supported by adaptive batching. The most important thing is that we need to choose more reasonable parameters to achieve a better performance.

Reply:
> @donzh , please provide command for TGI benchmark run

text-generation-benchmark --tokenizer-name /data/models--meta-llama--Meta-Llama-3-8B-Instruct/snapshots/e1945c40cd546c78e41f1151f4db032b271faeaa

Reply:
is that benchmark command for 8K total and 4K input ?

Reply:
> is that benchmark command for 8K total and 4K input ?

Yes, TGI needs to provide the 8K total capability.
Please set

#--max-input-tokens
export MAX_INPUT_TOKENS=4096
#--max-total-tokens
export MAX_TOTAL_TOKENS=8192

Reply:
> > @donzh I tried `meta-llama/Llama-2-7b-chat-hf` model with adaptive batching and I see this warning in logs, `2024-08-06T01:18:30.037184Z WARN text_generation_router: router/src/main.rs:393: Model does not support automatic max batch total tokens`
> > any specific reason this model doesn't support adaptive batching ?, which models are considered new models ? do we have list of models that support adaptive batching ?
> 
> This warning should not prevent you from starting the TGI server. If you don't use adaptive batching the warning exists. We test the Llama-3-8B, Mixtral and many other models. I think adaptive batching is not related with model. If a model is supported by TGI it should be supported by adaptive batching. The most important thing is that we need to choose more reasonable parameters to achieve a better performance.


Below is my command to start tgi-server and I see that warning. I am setting `MAX_TOTAL_TOKENS_2_MAX_BATCH_SIZE_LIST=2048:48,4096:24,8192:12`
```
docker run --rm -p 8080:80 \
   --runtime=habana \
   -v /software/users/thanaji/models:/data \
   -v /software/users/thanaji/models:/root/.cache/huggingface/hub \
   -v /home_local/labuser/thanaji/scripts/configs:/configs \
   -e QUANT_CONFIG=/configs/hqt_output_llama_7b_1card/maxabs_quant.json \
   -e HUGGING_FACE_HUB_TOKEN=<token> \
   -e HABANA_VISIBLE_DEVICES=7 \
   -e OMPI_MCA_btl_vader_single_copy_mechanism=none \
   -e HF_HUB_ENABLE_HF_TRANSFER=1 \
   -e PREFILL_BATCH_BUCKET_SIZE=1 \
   -e BATCH_BUCKET_SIZE=16 \
   -e PAD_SEQUENCE_TO_MULTIPLE_OF=128 \
   -e ENABLE_HPU_GRAPH=true \
   -e LIMIT_HPU_GRAPH=true \
   -e USE_FLASH_ATTENTION=true \
   -e FLASH_ATTENTION_RECOMPUTE=true \
   -e MAX_TOTAL_TOKENS_2_MAX_BATCH_SIZE_LIST=2048:48,4096:24,8192:12 \
   --cap-add=sys_nice \
   --ipc=host \
   thanaji_tgi_gaudi_adaptive:latest \
   --model-id meta-llama/Llama-2-7b-chat-hf \
   --max-input-tokens 1024 \
   --max-batch-prefill-tokens 4096 \
   --max-total-tokens 8192 \
   --max-batch-size 12
```

Reply:
@donzh in your previous comment you mentioned,  **If you _don't_ use adaptive batching the warning exists**. but we see that warning even when we use adaptive batching, could you explain or reword that warning better in which case the warning exists.

Reply:
This warning is not related with adaptive batching.